File: shoring/desk/src/main/java/com/panopset/marin/apps/fw/FxFunctionLoader.java
package com.panopset.marin.apps.fw;

import java.util.Collection;
import java.util.Collections;
import java.util.TreeSet;
import com.panopset.flywheel.Flywheel;
import com.panopset.flywheel.FlywheelFunction;
import com.panopset.flywheel.ReflectionInvoker;
import javafx.scene.control.ComboBox;
import javafx.scene.control.TextArea;

public class FxFunctionLoader {

  public void loadUpFunctions(ComboBox<FlywheelFunction> cbFunctions, TextArea fwTemplate) {
    populateDropdown(cbFunctions);
    wireUpEvents(cbFunctions, fwTemplate);
  }

  private void wireUpEvents(ComboBox<FlywheelFunction> cbFunctions, TextArea fwTemplate) {
    cbFunctions.setOnAction(event -> fwTemplate.appendText(cbFunctions.getValue().getExample()));
  }

  private void populateDropdown(ComboBox<FlywheelFunction> cbFunctions) {
    Flywheel.defineAllowedScriptCalls();
    cbFunctions.getItems().add(new FlywheelFunction());
    
    Collection<FlywheelFunction> allBut_dt = Collections.synchronizedSortedSet(new TreeSet<>());
    for (FlywheelFunction ff : ReflectionInvoker.getAll()) {
      if (!"dt".equalsIgnoreCase(ff.getKey())) {
        allBut_dt.add(ff);
      }
    }
    cbFunctions.getItems().addAll(allBut_dt);
  }
}

File: code/shoring/desk/src/main/kotlin/com/panopset/marin/apps/fw/FxFunctionLoader.kt
package com.panopset.marin.apps.fw

import com.panopset.flywheel.Flywheel.Companion.defineAllowedScriptCalls
import com.panopset.flywheel.FlywheelFunction
import com.panopset.flywheel.ReflectionInvoker.Companion.all
import javafx.event.EventHandler
import javafx.scene.control.ComboBox
import javafx.scene.control.TextArea
import java.util.*

class FxFunctionLoader {
    fun loadUpFunctions(cbFunctions: ComboBox<FlywheelFunction>, fwTemplate: TextArea) {
        populateDropdown(cbFunctions)
        wireUpEvents(cbFunctions, fwTemplate)
    }

    private fun wireUpEvents(cbFunctions: ComboBox<FlywheelFunction>, fwTemplate: TextArea) {
        cbFunctions.onAction = EventHandler { fwTemplate.appendText(cbFunctions.value.example) }
    }

    private fun populateDropdown(cbFunctions: ComboBox<FlywheelFunction>) {
        defineAllowedScriptCalls()
        cbFunctions.items.add(FlywheelFunction())

        val allBut_dt: MutableCollection<FlywheelFunction> = Collections.synchronizedSortedSet(TreeSet())
        for (ff in all) {
            if (!"dt".equals(ff.key, ignoreCase = true)) {
                allBut_dt.add(ff)
            }
        }
        cbFunctions.items.addAll(allBut_dt)
    }
}
